### PR TITLE
Gui_TinkoffInstrumentGraphic

### PR DESCRIPTION
**Pull request by ifelseelif**

### CODE REVIEW

File: Clients/GUI/Pages/_Host.cshtml [hunk 1]
```diff
@@ -8,32 +8,37 @@
...
+<script src="_content/BlazorInputFile/inputfile.js"></script>
+
+
+<script src="js/CandleChart.js"></script>
+<script src="https://cdn.jsdelivr.net/npm/apexcharts"></script>
 
-    <script src="_framework/blazor.server.js"></script>
-    <script src="_content/BlazorInputFile/inputfile.js"></script>
 </body>
-</html>
```
Comment (denis-kriv) [inline Clients/GUI/Pages/_Host.cshtml:39]: empty line

File: Clients/GUI/_Imports.razor [hunk 2]
```diff
@@ -33,3 +33,5 @@
 @using Kernel
 @using Kernel.Enums
 @using Kernel.CustomExceptions
+
+@using ChartJs.Blazor;
```
Comment (denis-kriv) [inline Clients/GUI/_Imports.razor:37]: empty line

File: Services/OperationService/BrokerConsumers/CandleConsumer.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        private readonly IHubContext<CandleHub> hubContext;
+
+        public CandleConsumer([FromServices] IHubContext<CandleHub> hubContext)
+        {
+            this.hubContext = hubContext;
+        }
+
+        public async Task Consume(ConsumeContext<Candle> context)
+        {
+            await hubContext.Clients.Group(context.Message.Figi).SendCoreAsync("ReceiveMessage", new[] {context.Message});
```
Comment (denis-kriv) [inline Services/OperationService/BrokerConsumers/CandleConsumer.cs]: May be line 21 it too long

Comment (ifelseelif) [reply Services/OperationService/BrokerConsumers/CandleConsumer.cs]: fixed

File: Clients/GUI/_Imports.razor (hunk 2 above)
Comment (ifelseelif) [reply Clients/GUI/_Imports.razor:37]: fixed

File: Clients/GUI/Pages/_Host.cshtml (hunk 1 above)
Comment (ifelseelif) [reply Clients/GUI/Pages/_Host.cshtml:39]: fixed

File: Clients/GUI/Pages/InstrumentGraphic.razor [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            token = await Storage.GetItemAsync<string>("tinkoffToken");
+            if (token == null)
+            {
+                errorMessage = "Access error";
+                StateHasChanged();
+                return;
+            }
+
+            hubConnection = new HubConnectionBuilder()
```
Comment (dTarasova) [inline Clients/GUI/Pages/InstrumentGraphic.razor]: Not sure, but it wouldn't be better to connect to hub in the other class? In razor page it is not the best way to see logic

**Review by dTarasova — COMMENTED** [review]
Graphic seems like just a line. I see how do I can regulate date (but I can't move them, as I understood, but I can't regulate depth and to see small changes in the price (probably I couldn't find how to do it

Diign seems good, but probably we have to make graphic bigger. A lot of white space. Like the opportunity to save stuff

image

File: Clients/GUI/Pages/InstrumentGraphic.razor (hunk 4 above)
Comment (ifelseelif) [reply Clients/GUI/Pages/InstrumentGraphic.razor]: graphic changed. Hub moved in the other class

**Review by Cerberous007 — APPROVED** [review]

File: Services/OperationService/Hubs/CandleHub.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public class CandleHub : Hub
+    {
+        private readonly IRequestClient<SubscribeOnCandleRequest> client;
+
+        public CandleHub([FromServices] IRequestClient<SubscribeOnCandleRequest> client)
+        {
+            this.client = client;
+        }
+        public async Task Subscribe(SubscribeOnCandleRequest request)
```
Comment (nmamro) [inline Services/OperationService/Hubs/CandleHub.cs]: Line

File: Services/OperationService/Hubs/CandleHub.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        {
+            await Groups.AddToGroupAsync(Context.ConnectionId,request.Figi);
+            await client.GetResponse<OperationResult>(request);
+
```
Comment (nmamro) [inline Services/OperationService/Hubs/CandleHub.cs]: Empty

File: Clients/GUI/Pages/InstrumentGraphic.razor [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+@using ChartJs.Blazor.ChartJS.Common.Properties
+@using ChartJs.Blazor.ChartJS.Common.Time
+@using ChartJs.Blazor.ChartJS.LineChart
+@using ChartJs.Blazor.Charts
+@using GUI.Scripts
+@inject NavigationManager NavigationManager
+@inject ISessionStorageService Storage
+@inject IJSRuntime JsRuntime
+
+<ChartJsLineChart @ref="lineChartJs" Config="@lineConfig" Width="600" Height="300"/>
```
Comment (nmamro) [inline Clients/GUI/Pages/InstrumentGraphic.razor:16]: Can you make the graph a bit brighter? (light colour blends with white background a bit)

File: Clients/GUI/Pages/InstrumentGraphic.razor [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+@code {
+
+    [Parameter]
+    public string Figi { get; set; }
+
+    private string Name = "";
```
Comment (nmamro) [inline Clients/GUI/Pages/InstrumentGraphic.razor]: Why capital N?

File: Clients/GUI/Pages/InstrumentGraphic.razor [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    private string token = "";
+    private ChartJsLineChart lineChartJs;
+    private LineConfig lineConfig;
+    private LineDataset<TimeTuple<decimal>> tempDataSet;
+    private DateTime lasTime;
```
Comment (nmamro) [inline Clients/GUI/Pages/InstrumentGraphic.razor]: las?

File: Clients/GUI/Pages/InstrumentGraphic.razor [hunk 10]
```diff
@@ -0,0 +1,143 @@
+﻿@page "/instrument/{Figi}"
+@using ChartJs.Blazor.ChartJS.Common.Axes
+@using ChartJs.Blazor.ChartJS.Common.Axes.Ticks
+@using ChartJs.Blazor.ChartJS.Common.Enums
+@using ChartJs.Blazor.ChartJS.Common.Handlers
+@using ChartJs.Blazor.ChartJS.Common.Properties
+@using ChartJs.Blazor.ChartJS.Common.Time
+@using ChartJs.Blazor.ChartJS.LineChart
+@using ChartJs.Blazor.Charts
+@using GUI.Scripts
```
Comment (nmamro) [inline Clients/GUI/Pages/InstrumentGraphic.razor:12]: Remove it from here

File: Clients/GUI/Pages/InstrumentGraphic.razor [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+@inject NavigationManager NavigationManager
+@inject ISessionStorageService Storage
+@inject IJSRuntime JsRuntime
```
Comment (nmamro) [inline Clients/GUI/Pages/InstrumentGraphic.razor]: Do you use it?

File: Clients/GUI/Pages/InstrumentGraphic.razor [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+@using ChartJs.Blazor.ChartJS.Common.Axes
+@using ChartJs.Blazor.ChartJS.Common.Axes.Ticks
+@using ChartJs.Blazor.ChartJS.Common.Enums
+@using ChartJs.Blazor.ChartJS.Common.Handlers
+@using ChartJs.Blazor.ChartJS.Common.Properties
+@using ChartJs.Blazor.ChartJS.Common.Time
+@using ChartJs.Blazor.ChartJS.LineChart
+@using ChartJs.Blazor.Charts
+@using GUI.Scripts
+@inject NavigationManager NavigationManager
```
Comment (nmamro) [inline Clients/GUI/Pages/InstrumentGraphic.razor:13]: Do you use this one?

File: Common/DTO/Candle.cs [hunk 13]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+
+namespace DTO
+{
+    public class Candle
```
Comment (nmamro) [inline Common/DTO/Candle.cs:5]: I'm not sure if it should still be called candle as chart is no longer a candle chat, but ok

File: Integrations/TinkoffIntegrationLib/TinkoffBankBroker.cs [hunk 14]
```diff
@@ -34,39 +37,57 @@ public TinkoffBankBroker(string token, int depth = 10)
...
                     {
                         instruments.Add(
                             new TinkoffInstrumentAdapter(
                                 tinkoffInstrumentType,
                                 instrument)
-                            );
+                        );
+                    }
+                    catch (Exception)
+                    {
```
Comment (nmamro) [inline Integrations/TinkoffIntegrationLib/TinkoffBankBroker.cs]: No idea why you reformatted it

File: Integrations/TinkoffIntegrationLib/TinkoffBankBroker.cs [hunk 15]
```diff
@@ -34,39 +37,57 @@ public TinkoffBankBroker(string token, int depth = 10)
...
+
+        public void SubscribeOnCandle(string Figi, Action<Candle> SendCandle)
+        {
+            this.sendCandle = SendCandle;
+            context.StreamingEventReceived += onStreamingEventReceived;
+            context.SendStreamingRequestAsync(new StreamingRequest.CandleSubscribeRequest(Figi, CandleInterval.Minute));
+        }
+
+
+        private void onStreamingEventReceived(object sender, StreamingEventReceivedEventArgs args)
```
Comment (nmamro) [inline Integrations/TinkoffIntegrationLib/TinkoffBankBroker.cs]: Methods start with capitals

**Review by nmamro — CHANGES_REQUESTED** [review]
Seems ok overall

File: Integrations/TinkoffIntegrationLib/TinkoffBankBroker.cs (hunk 15 above)
Comment (ifelseelif) [reply Integrations/TinkoffIntegrationLib/TinkoffBankBroker.cs]: fixed

File: Integrations/TinkoffIntegrationLib/TinkoffBankBroker.cs (hunk 14 above)
Comment (ifelseelif) [reply Integrations/TinkoffIntegrationLib/TinkoffBankBroker.cs]: miss click

File: Clients/GUI/Pages/InstrumentGraphic.razor (hunk 12 above)
Comment (ifelseelif) [reply Clients/GUI/Pages/InstrumentGraphic.razor:13]: deleted

File: Clients/GUI/Pages/InstrumentGraphic.razor (hunk 11 above)
Comment (ifelseelif) [reply Clients/GUI/Pages/InstrumentGraphic.razor]: deleted

File: Clients/GUI/Pages/InstrumentGraphic.razor (hunk 7 above)
Comment (ifelseelif) [reply Clients/GUI/Pages/InstrumentGraphic.razor:16]: Color changed on blue

**Review by dTarasova — APPROVED** [review]

**Review by denis-kriv — APPROVED** [review]

**Review by Maksim-Trolina — APPROVED** [review]

**Review by strannick-rodion — APPROVED** [review]

**Review by nmamro — APPROVED** [review]

**Pull request merged**